### PR TITLE
AIP-4600: [Part 3 of 3] Pull in upstream fixes from KFP.

### PR DESCRIPTION
**Pull request by alexlatchford**
Parts:
1. https://github.com/zillow/pipelines/pull/7
2. https://github.com/zillow/pipelines/pull/8
3. https://github.com/zillow/pipelines/pull/9 - This PR

Pulls in these fixes we contributed upstream to KFP:
* https://github.com/kubeflow/pipelines/pull/6504
* https://github.com/kubeflow/pipelines/pull/6532

### CODE REVIEW

File: sdk/python/kfp/components/_data_passing.py [hunk 1]
```diff
@@ -139,23 +139,24 @@ def get_canonical_type_for_type_struct(type_struct) -> str:
         return None
 
 
-def get_deserializer_code_for_type(typ) -> str:
-    try:
-        return type_name_to_deserializer.get(get_canonical_type_struct_for_type[typ], None)
-    except:
-        return None
+# Doesn't seem to be used anymore, can be removed?
```
Comment (hsezhiyan) [inline sdk/python/kfp/components/_data_passing.py:142]: Can we just delete this unused code? Makes the repo much cleaner I'd say.

File: sdk/python/kfp/components/_data_passing.py [hunk 2]
```diff
@@ -139,23 +139,24 @@ def get_canonical_type_for_type_struct(type_struct) -> str:
...
+#         return None
 
 
-def get_deserializer_code_for_type_struct(type_struct) -> str:
+def get_deserializer_code_for_type_struct(type_struct: str) -> str:
     try:
-        return type_name_to_deserializer.get(type_struct, None)
+        return type_name_to_deserializer.get(type_annotation_utils.get_short_type_name(type_struct), None)
     except:
         return None
```
Comment (hsezhiyan) [inline sdk/python/kfp/components/_data_passing.py:154]: Hmm, will this `except` clause ever be reached? In the `try` class, you do a `get` and provide a default `None`, so it should never raise an exception?

File: sdk/python/kfp/components/_data_passing.py (hunk 1 above)
Comment (alexlatchford) [reply sdk/python/kfp/components/_data_passing.py:142]: Yeah I made a separate branch for contributing to `kfp` upstream, see [here](https://github.com/kubeflow/pipelines/pull/6504) and I got the go ahead to remove this. I'll actually be closing this PR and just be `cherry-picking` in the upstream change once it get's merged (hopefully today!).

File: sdk/python/kfp/components/_data_passing.py (hunk 2 above)
Comment (alexlatchford) [reply sdk/python/kfp/components/_data_passing.py:154]: Yeah good question, it get's triggered when `type_struct` is not a `str`. Not ideal as ideally the caller of these methods are conforming to the designated `type` of the input arguments but to be honest I didn't fancy trying to fix all the references to this across `kfp` so I just kept the `except` in place 😄

Comment (hsezhiyan) [reply sdk/python/kfp/components/_data_passing.py:170]: Ahh makes sense, thanks for letting me know!

**Review by hsezhiyan — APPROVED** [review]

**Pull request merged**